core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -762,14 +763,14 @@ public ReplicationPhase(ReplicaRequest replicaRequest, Response finalResponse, S
             // we have to make sure that every operation indexed into the primary after recovery start will also be replicated
             // to the recovery target. If we use an old cluster state, we may miss a relocation that has started since then.
             // If the index gets deleted after primary operation, we skip replication
-            final ClusterState state = clusterService.state();
-            final IndexRoutingTable index = state.getRoutingTable().index(shardId.getIndex());
+            this.clusterState = clusterService.state();
```

Comment:
I rather not cache the cluster state - it can be big (problems in the past). That's why we moved to just using the nodes and shards. It seems we need it get the primary shard routing. I'd rather have it from the indexShardReference , which indicates the primary that actually does the job. Se for example: https://github.com/elastic/elasticsearch/pull/15900/files#diff-a8aefbf42f29dc0fcc7c0a144863948eR1014 .  I hope that PR can go in soon so you can reuse that infra.